torchao/dtypes/nf4tensor.py

```diff
@@ -562,7 +858,10 @@ class LinearNF4(torch.autograd.Function):
     def forward(ctx, input: torch.Tensor, weight: NF4Tensor):
         """Save the quantized nf4 weight for backward pass"""
         ctx.nf4_weight = weight
-        return F.linear(input, weight.to(input.dtype))
+        # assert input.dtype == torch.bfloat16 and input.dtype == weight.dtype
+        assert input.dtype == weight.dtype
+        return F.linear(input, weight.get_original_weight())
```

Comment:
@cpuhrsch since `get_original_weight` returns original constructor dtype (fp32/fp16/bf16), I removed `weight.to` and diabled `torch.ops.aten.t.default` in dispatching. is this making sense?

Reply:
keep using `weight.to` after implementing `to` in `__torch_function__`